### PR TITLE
[InstSimplify] Add simplification for `({u,s}rem (mul {nuw,nsw} X, C1), C0)`

### PR DESCRIPTION
**Pull request by goldsteinn**
- **[InstSimplify] Add test for simplifying `({u,s}rem (mul {nuw,nsw} X, C1), C0)`; NFC**
- **[InstSimplify] Add simplification for `({u,s}rem (mul {nuw,nsw} X, C1), C0)`**

We can simplify these to `0` if `C1 % C0 == 0`

Proofs: https://alive2.llvm.org/ce/z/EejAdk

### CODE REVIEW

File: llvm/lib/Analysis/InstructionSimplify.cpp [hunk 1]
```diff
@@ -1233,6 +1233,19 @@ static Value *simplifyRem(Instruction::BinaryOps Opcode, Value *Op0, Value *Op1,
         match(Op0, m_NUWShl(m_Specific(Op1), m_Value())))))
     return Constant::getNullValue(Op0->getType());
 
+  const APInt * C0;
+  if (match(Op1, m_APInt(C0))) {
```
Comment (nikic) [inline llvm/lib/Analysis/InstructionSimplify.cpp:1237]: Missing Q.IIQ.UseInstrInfo check.

File: llvm/lib/Analysis/InstructionSimplify.cpp [hunk 2]
```diff
@@ -1233,6 +1233,19 @@ static Value *simplifyRem(Instruction::BinaryOps Opcode, Value *Op0, Value *Op1,
         match(Op0, m_NUWShl(m_Specific(Op1), m_Value())))))
     return Constant::getNullValue(Op0->getType());
 
+  const APInt * C0;
```
Comment (nikic) [inline llvm/lib/Analysis/InstructionSimplify.cpp:1236]:
```suggestion
  const APInt *C0;
```

File: llvm/test/Transforms/InstSimplify/rem.ll [hunk 3]
```diff
@@ -265,12 +265,14 @@ define i32 @rem4() {
...
-  %call = call i32 @external(), !range !0
+  %call = call i32 @external() , !range !0
   %urem = urem i32 %call, 3
   ret i32 %urem
 }
 
-!0 = !{i32 0, i32 3}
+!0 = ! {
+i32 0, i32 3
+}
```
Comment (nikic) [inline llvm/test/Transforms/InstSimplify/rem.ll:275]: Something went wrong here?

**Review by nikic — APPROVED** [review]
LGTM, surprised we didn't have this already...

File: llvm/test/Transforms/InstSimplify/rem.ll [hunk 4]
```diff
@@ -488,3 +490,54 @@ define i8 @urem_mul_sdiv(i8 %x, i8 %y) {
...
+  ret i8 %r
+}
+
+define i8 @simplfy_urem_of_mul_fail_bad_mod(i8 %x) {
+; CHECK-LABEL: @simplfy_urem_of_mul_fail_bad_mod(
+; CHECK-NEXT:    [[MUL:%.*]] = mul nsw i8 [[X:%.*]], 31
+; CHECK-NEXT:    [[R:%.*]] = urem i8 [[MUL]], 10
+; CHECK-NEXT:    ret i8 [[R]]
+;
+  %mul = mul nsw i8 %x, 31
```
Comment (nikic) [inline llvm/test/Transforms/InstSimplify/rem.ll:540]:
```suggestion
  %mul = mul nuw i8 %x, 31
```
Otherwise this is not a "bad_mod" test.

**Review by dtcxzyw — APPROVED** [review]
LGTM.

**Pull request closed without merging**